src/ext/php5_4/engine_hooks.c

```diff
@@ -1,550 +1,655 @@
+    if (keep_span) {
+        dd_set_default_properties(span_fci, fbc TSRMLS_CC);
+        ddtrace_close_span(TSRMLS_C);
+    } else {
+        ddtrace_drop_top_open_span(TSRMLS_C);
+    }
+}
 
-    // Move this to closure zval before zend_fcall_info_init()
-    fcc.function_handler->common.function_name = func_name;
+static void dd_execute_tracing_posthook(zend_op_array *op_array TSRMLS_DC) {
+    ddtrace_dispatch_t *dispatch = op_array->reserved[ddtrace_resource];
+    zend_function *fbc = (zend_function *)op_array;
 
-    zend_execute_data *prev_original_execute_data = DDTRACE_G(original_context).execute_data;
-    DDTRACE_G(original_context).execute_data = execute_data;
+    bool free_retval = false;
+    zval *retval = NULL;
 
-    zval *prev_original_function_name = DDTRACE_G(original_context).function_name;
-    DDTRACE_G(original_context).function_name = (*EG(opline_ptr))->op1.zv;
+    /* If the retval doesn't get used then sometimes the engine won't set
+     * the retval_ptr_ptr at all. We expect it to always be present, so
+     * adjust it. Be sure to dtor it later.
+     */
+    if (!EG(return_value_ptr_ptr)) {
+        EG(return_value_ptr_ptr) = &retval;
+        free_retval = true;
+    }
 
-    zend_call_function(&fci, &fcc TSRMLS_CC);
+    ddtrace_dispatch_copy(dispatch);
 
-    DDTRACE_G(original_context).function_name = prev_original_function_name;
+    ddtrace_span_fci *span_fci = ecalloc(1, sizeof(ddtrace_span_fci));
+    span_fci->dispatch = dispatch;
+    ddtrace_open_span(span_fci TSRMLS_CC);
 
-    DDTRACE_G(original_context).execute_data = prev_original_execute_data;
+    dd_prev_execute(op_array TSRMLS_CC);
 
-    if (fci.params) {
-        efree(fci.params);
-    }
+    /* Sometimes the retval goes away when there is an exception, and
+     * sometimes it's there but points to nothing (even excluding our fixup
+     * above), so check both.
+     */
+    zval *actual_retval =
+        (EG(return_value_ptr_ptr) && *EG(return_value_ptr_ptr)) ? *EG(return_value_ptr_ptr) : &zval_used_for_init;
 
-_exit_cleanup:
+    if (span_fci == DDTRACE_G(open_spans_top)) {
+        dd_tracing_posthook_impl(fbc, span_fci, actual_retval TSRMLS_CC);
+    } else {
+        if (get_dd_trace_debug()) {
+            // todo: update to Classname::Methodname format
+            const char *fname = Z_STRVAL(dispatch->function_name);
+            ddtrace_log_errf("Cannot run tracing closure for %s(); spans out of sync", fname);
+        }
+    }
 
-    if (this) {
-        Z_DELREF_P(this);
+    if (free_retval && *EG(return_value_ptr_ptr)) {
+        zval_ptr_dtor(EG(return_value_ptr_ptr));
+        EG(return_value_ptr_ptr) = NULL;
     }
-    Z_DELREF(closure);
-    zval_dtor(&closure);
-    DDTRACE_G(original_context).fbc = current_fbc;
 }
 
-static void wrap_and_run(zend_execute_data *execute_data, ddtrace_dispatch_t *dispatch TSRMLS_DC) {
-    zval *this = ddtrace_this(execute_data);
+// copy dispatch on outside!
```

Comment:
Was this a note to self? :)

Reply:
It's documentation! Apparently not very good :)